alacritty/src/renderer/text/gles2.rs

```diff
@@ -0,0 +1,494 @@
+
+            // Color and bitmap color.
+            //
+            // These are packed together because of an OpenGL driver issue on macOS, which caused a
+            // `vec3(u8)` text color and a `u8` for glyph color to cause performance regressions.
+            add_attr!(4, gl::UNSIGNED_BYTE, u8);
+
+            // Background color.
+            add_attr!(4, gl::UNSIGNED_BYTE, u8);
+
+            // Cleanup.
+            gl::BindVertexArray(0);
+            gl::BindBuffer(gl::ELEMENT_ARRAY_BUFFER, 0);
+            gl::BindBuffer(gl::ARRAY_BUFFER, 0);
+        }
+
+        Ok(Self {
+            program,
+            vao,
+            vbo,
+            ebo,
+            atlas: vec![Atlas::new(ATLAS_SIZE)],
+            batch: Batch::new(),
+            current_atlas: 0,
+            active_tex: 0,
+        })
+    }
+
+    fn with_api<F, T>(&mut self, _: &SizeInfo, func: F) -> T
+    where
+        F: FnOnce(RenderApi<'_>) -> T,
+    {
+        unsafe {
+            gl::UseProgram(self.program.id());
+            gl::BindVertexArray(self.vao);
+            gl::BindBuffer(gl::ELEMENT_ARRAY_BUFFER, self.ebo);
+            gl::BindBuffer(gl::ARRAY_BUFFER, self.vbo);
+            gl::ActiveTexture(gl::TEXTURE0);
+        }
+
+        let res = func(RenderApi {
+            active_tex: &mut self.active_tex,
+            batch: &mut self.batch,
+            atlas: &mut self.atlas,
+            current_atlas: &mut self.current_atlas,
+            program: &mut self.program,
+        });
+
+        unsafe {
+            gl::BindBuffer(gl::ELEMENT_ARRAY_BUFFER, 0);
+            gl::BindBuffer(gl::ARRAY_BUFFER, 0);
+            gl::BindVertexArray(0);
+
+            gl::UseProgram(0);
+        }
+
+        res
+    }
+}
+
+impl Drop for Gles2Renderer {
+    fn drop(&mut self) {
+        unsafe {
+            gl::DeleteBuffers(1, &self.vbo);
+            gl::DeleteBuffers(1, &self.ebo);
+            gl::DeleteVertexArrays(1, &self.vao);
+        }
+    }
+}
+
+impl TextRenderer for Gles2Renderer {
+    fn draw_cells<I: Iterator<Item = RenderableCell>>(
+        &mut self,
+        size_info: &SizeInfo,
+        glyph_cache: &mut GlyphCache,
+        cells: I,
+    ) {
+        self.with_api(size_info, |mut api| {
+            for cell in cells {
+                api.draw_cell(cell, glyph_cache, size_info)
+            }
+        })
+    }
+
+    fn resize(&self, size_info: &SizeInfo) {
+        unsafe {
+            gl::UseProgram(self.program.id());
+            self.program.update_projection(
+                size_info.width(),
+                size_info.height(),
+                size_info.padding_x(),
+                size_info.padding_y(),
+            );
+            gl::UseProgram(0);
+        }
+    }
+
+    fn loader_api(&mut self) -> LoaderApi<'_> {
+        LoaderApi {
+            active_tex: &mut self.active_tex,
+            atlas: &mut self.atlas,
+            current_atlas: &mut self.current_atlas,
+        }
+    }
+}
```

Comment:
This trait implementation is basically exactly the same for both renderers, which is usually a sign some other abstraction might be beneficial?

Instead of the `resize` method, you could use a `program` method for example which would allow you to just have one line duplicated which would be all safe code too.

The `draw_cells` could maybe be abstracted to `with_api`? The function exists already anyway and at that point we'd just define `draw_cells` once somewhere instead of doing it twice.